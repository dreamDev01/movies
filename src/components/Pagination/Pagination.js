import React, { Component } from "react";

import "./Pagination.scss";

/*-------Pagination----------*/

class Pagination extends Component {
  /*Go to the selected page*/
  onPage(n) {
    this.props.onPage(n);
  }

  /*is the last page or not*/
  isLastPage() {
    return (
      this.props.perPage * this.props.currentPage >= this.props.totalItemsCount
    );
  }

  /*total page*/
  total() {
    return Math.ceil(this.props.totalItemsCount / this.props.perPage) || 0;
  }

  /*get minimum page*/
  getMin() {
    return this.props.perPage * this.props.currentPage - this.props.perPage + 1;
  }

  /*get maximum page*/
  getMax() {
    let max = this.props.perPage * this.props.currentPage;
    if (max > this.props.totalItemsCount) {
      max = this.props.totalItemsCount;
    }
    return max;
  }

  /*go to the previous page*/
  onPrevious = () => {
    this.props.onPrevPage();
  };

  /*go to the next page*/
  onNext = () => {
    this.props.onNextPage();
  };

  /*get all pages*/
  getAllPages = () => {
    const c = Math.ceil(this.props.totalItemsCount / this.props.perPage);
    const p = this.props.currentPage || 1;
    const pagesToShow = this.props.pagesToShow || 9;
    const pages = [];
    pages.push(p);
    const times = pagesToShow - 1;
    for (let i = 0; i < times; i++) {
      if (pages.length < pagesToShow) {
        if (Math.min.apply(null, pages) > 1) {
          pages.push(Math.min.apply(null, pages) - 1);
        }
      }
      if (pages.length < pagesToShow) {
        if (Math.max.apply(null, pages) < c) {
          pages.push(Math.max.apply(null, pages) + 1);
        }
      }
    }
    pages.sort((a, b) => a - b);
    return pages;
  };

  render() {
    const pages = this.getAllPages().map((pageNum, index) => {
      let buttonClass = "page-item";

      if (pageNum === this.props.currentPage) {
        buttonClass += " active";
      }

      return (
        <li
          key={index}
          className={buttonClass}
          onClick={() => {
            this.onPage(pageNum);
          }}
        >
          <button className="page-link">{pageNum}</button>
        </li>
      );
    });

    let prevButtonClass = "page-item";

    if (this.props.currentPage === 1) {
      prevButtonClass += " disabled";
    }

    const prevButton = (
      <li className={prevButtonClass}>
        <button
          data-testid="prev"
          className="page-link"
          onClick={this.onPrevious}
          tabIndex="-1"
        >
          Previous
        </button>
      </li>
    );

    let nextButtonClass = "page-item";

    if (this.isLastPage()) {
      nextButtonClass += " disabled";
    }

    const nextButton = (
      <li className={nextButtonClass}>
        <button
          data-testid="next"
          disabled={this.isLastPage()}
          className="page-link"
          onClick={this.onNext}
        >
          Next
        </button>
      </li>
    );

    return (
      <nav aria-label="...">
        <ul className="pagination">
          {prevButton}
          {pages}
          {nextButton}
        </ul>
      </nav>
    );
  }
}

export default Pagination;
