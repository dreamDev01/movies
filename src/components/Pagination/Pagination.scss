.pagination {
  display: flex;
  ul {
    list-style-type: none;
  }
}

.page-link {
  position: relative;
  display: block;
  padding: 0.5rem 0.75rem;
  margin-left: -1px;
  line-height: 1.25;
  color: #007bff;
  background-color: #fff;
  border: 1px solid #dee2e6;
  cursor: pointer;

  &:hover {
    z-index: 2;
    text-decoration: none;
  }

  &:focus {
    z-index: 2;
  }
}

.page-item {
  &:first-child {
    .page-link {
      margin-left: 0;
      border-top-left-radius: 0.25rem;
      border-bottom-left-radius: 0.25rem;
    }
  }
  &:last-child {
    .page-link {
      border-top-right-radius: 0.25rem;
      border-bottom-right-radius: 0.25rem;
    }
  }

  &.active .page-link {
    z-index: 1;
    color: #fff;
    background-color: #007bff;
    border-color: #007bff;
  }

  &.disabled .page-link {
    color: #6c757d;
    pointer-events: none;
    // Opinionated: remove the "hand" cursor set previously for .page-link
    cursor: auto;
    background-color: #fff;
    border-color: #e9ecef;
  }
}
