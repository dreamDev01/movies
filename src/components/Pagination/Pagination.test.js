import renderer from "react-test-renderer";
import { render, screen } from "@testing-library/react";

import Pagination from "./Pagination";

describe("tests Pagination", () => {
  it("should render component", async () => {
    // Arrange
    const tree = renderer.create(<Pagination />).toJSON();

    // Assert
    expect(tree).toMatchSnapshot();
  });

  describe("display", () => {
    it("should display pagination", async () => {
      // Act

      render(<Pagination />);

      // Assert
      expect(screen.getByTestId("next").textContent).toEqual("Next");
      expect(screen.getByTestId("prev").textContent).toEqual("Previous");
    });
  });
});
