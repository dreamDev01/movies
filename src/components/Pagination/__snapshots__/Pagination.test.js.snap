// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests Pagination should render component 1`] = `
<nav
  aria-label="..."
>
  <ul
    className="pagination"
  >
    <li
      className="page-item"
    >
      <button
        className="page-link"
        data-testid="prev"
        onClick={[Function]}
        tabIndex="-1"
      >
        Previous
      </button>
    </li>
    <li
      className="page-item"
      onClick={[Function]}
    >
      <button
        className="page-link"
      >
        1
      </button>
    </li>
    <li
      className="page-item"
    >
      <button
        className="page-link"
        data-testid="next"
        disabled={false}
        onClick={[Function]}
      >
        Next
      </button>
    </li>
  </ul>
</nav>
`;
